src/r1cs/sparse.rs

```diff
@@ -0,0 +1,249 @@
+//! # Sparse Matrices
+//!
+//! This module defines a custom implementation of CSR/CSC sparse matrices.
+//! Specifically, we implement sparse matrix / dense vector multiplication
+//! to compute the `A z`, `B z`, and `C z` in Nova.
+
+use abomonation::Abomonation;
+use abomonation_derive::Abomonation;
+use ff::PrimeField;
+use rayon::prelude::*;
+use serde::{Deserialize, Serialize};
+
+/// CSR format sparse matrix, We follow the names used by scipy.
+/// Detailed explanation here: https://stackoverflow.com/questions/52299420/scipy-csr-matrix-understand-indptr
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Abomonation)]
+#[abomonation_bounds(where <F as PrimeField>::Repr: Abomonation)]
+pub struct SparseMatrix<F: PrimeField> {
+  #[abomonate_with(Vec<F::Repr>)]
+  /// all non-zero values in the matrix
+  pub data: Vec<F>,
+  /// column indices
+  pub indices: Vec<usize>,
+  /// row information
+  pub indptr: Vec<usize>,
+  /// number of columns
+  pub cols: usize,
+}
+
+impl<F: PrimeField> SparseMatrix<F> {
+  /// 0x0 empty matrix
+  pub fn empty() -> Self {
+    SparseMatrix {
+      data: vec![],
+      indices: vec![],
+      indptr: vec![0],
+      cols: 0,
+    }
+  }
+
+  /// construct from Vec<usize(row), usize(col), F>
+  pub fn new(matrix: &[(usize, usize, F)], rows: usize, cols: usize) -> Self {
```

Comment:
Is this needed for anything but unit testing? If not, can we assume at least that unit testing will provide matrices in sorted form, and then remove any correspondingly superfluous methods? If unit tests don't use this, would this be an invitation to?

Reply:
I've removed the sorting methods, but I've kept `::new()` for now, as it's used in `test_tiny_r1cs_with`. Also I've found that debugging is easier in COO form, so having a way to transform from COO to CSR via `::new()` and CSR to COO via `iter()` is nice.

Reply:
Feel free to add the sorted assumption in `new`'s comment.